elytron/src/main/java/org/wildfly/extension/elytron/VirtualDomainDefinition.java

```diff
@@ -0,0 +1,207 @@
+import static org.wildfly.extension.elytron.DomainDefinition.outflow;
+import static org.wildfly.extension.elytron.ElytronDefinition.commonDependencies;
+import static org.wildfly.extension.elytron.ElytronDescriptionConstants.INITIAL;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.UnaryOperator;
+
+import org.jboss.as.controller.AbstractAddStepHandler;
+import org.jboss.as.controller.AttributeDefinition;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.OperationStepHandler;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.PathElement;
+import org.jboss.as.controller.ResourceDefinition;
+import org.jboss.as.controller.SimpleResourceDefinition;
+import org.jboss.as.controller.StringListAttributeDefinition;
+import org.jboss.as.controller.capability.RuntimeCapability;
+import org.jboss.as.controller.registry.AttributeAccess;
+import org.jboss.as.controller.registry.ManagementResourceRegistration;
+import org.jboss.as.controller.registry.OperationEntry;
+import org.jboss.as.server.security.VirtualDomainMetaData;
+import org.jboss.as.server.security.VirtualDomainMetaDataService;
+import org.jboss.dmr.ModelNode;
+import org.jboss.msc.service.ServiceBuilder;
+import org.jboss.msc.service.ServiceController;
+import org.jboss.msc.service.ServiceController.Mode;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.ServiceTarget;
+import org.jboss.msc.service.StartException;
+import org.jboss.msc.value.InjectedValue;
+import org.wildfly.extension.elytron.TrivialService.ValueSupplier;
+import org.wildfly.security.auth.server.SecurityDomain;
+import org.wildfly.security.auth.server.SecurityIdentity;
+
+/**
+ * A {@link ResourceDefinition} for a virtual security domain.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+class VirtualDomainDefinition extends SimpleResourceDefinition {
+
+    static final StringListAttributeDefinition OUTFLOW_SECURITY_DOMAINS = new StringListAttributeDefinition.Builder(ElytronDescriptionConstants.OUTFLOW_SECURITY_DOMAINS)
+            .setRequired(false)
+            .setMinSize(1)
+            .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+            .setCapabilityReference(SECURITY_DOMAIN_CAPABILITY, VIRTUAL_SECURITY_DOMAIN_CAPABILITY)
+            .build();
+
+    private static final AttributeDefinition[] ATTRIBUTES = new AttributeDefinition[] { OUTFLOW_SECURITY_DOMAINS, OUTFLOW_ANONYMOUS };
+
+    private static final VirtualDomainAddHandler ADD = new VirtualDomainAddHandler();
+    private static final OperationStepHandler REMOVE = new VirtualDomainRemoveHandler(ADD);
+    private static final WriteAttributeHandler WRITE = new WriteAttributeHandler(ElytronDescriptionConstants.VIRTUAL_SECURITY_DOMAIN);
```

Comment:
When adding `virtual-security-domain` with `outflow-security-domains`, the resource is created as expected. When adding an empty `virtual-security-domain` and writing `outflow-security-domains`, an error occurs:

```
/subsystem=elytron/virtual-security-domain=abcde.war:add(outflow-security-domains=[ApplicationDomain,ManagementDomain])
{"outcome" => "success"}


/subsystem=elytron/virtual-security-domain=abc.war:add()
/subsystem=elytron/virtual-security-domain=abc.war:write-attribute(name=outflow-security-domains,value=[ApplicationDomain,ManagementDomain])
{
    "outcome" => "failed",
    "failure-description" => "WFLYCTL0158: Operation handler failed: java.lang.IllegalArgumentException: WFLYCTL0394: Capability 'org.wildfly.security.virtual-security-domain.abc.war' does not provide services of type 'class org.wildfly.security.auth.server.SecurityDomain'",
    "rolled-back" => true
}
```

Reply:
Good catch, fixed.